types/query/pagination_test.go

```diff
@@ -0,0 +1,199 @@
+	request = types.NewQueryAllBalancesRequest(addr1, pageReq)
+	res, err = queryClient.AllBalances(gocontext.Background(), request)
+	require.NoError(t, err)
+	require.LessOrEqual(t, res.Balances.Len(), defaultLimit)
+	require.Equal(t, res.Balances.Len(), lastPageRecords)
+	require.Nil(t, res.Res.NextKey)
+	require.Equal(t, res.Res.Total, uint64(0))
+
+	t.Log("verify paginate with offset and limit")
+	pageReq = &query.PageRequest{Offset: 200, Limit: defaultLimit, CountTotal: false}
+	request = types.NewQueryAllBalancesRequest(addr1, pageReq)
+	res, err = queryClient.AllBalances(gocontext.Background(), request)
+	require.NoError(t, err)
+	require.LessOrEqual(t, res.Balances.Len(), defaultLimit)
+	require.Equal(t, res.Balances.Len(), lastPageRecords)
+	require.Nil(t, res.Res.NextKey)
+	require.Equal(t, res.Res.Total, uint64(0))
+
+	t.Log("verify paginate with offset and limit")
+	pageReq = &query.PageRequest{Offset: 100, Limit: defaultLimit, CountTotal: false}
+	request = types.NewQueryAllBalancesRequest(addr1, pageReq)
+	res, err = queryClient.AllBalances(gocontext.Background(), request)
+	require.NoError(t, err)
+	require.LessOrEqual(t, res.Balances.Len(), defaultLimit)
+	require.NotNil(t, res.Res.NextKey)
+	require.Equal(t, res.Res.Total, uint64(0))
+
+	t.Log("verify paginate with offset and key - error")
+	pageReq = &query.PageRequest{Key: res.Res.NextKey, Offset: 100, Limit: defaultLimit, CountTotal: false}
+	request = types.NewQueryAllBalancesRequest(addr1, pageReq)
+	res, err = queryClient.AllBalances(gocontext.Background(), request)
+	require.Error(t, err)
+	require.Equal(t, err.Error(), "invalid request, either offset or key is expected, got both")
+
+	t.Log("verify paginate with offset greater than total results")
+	pageReq = &query.PageRequest{Offset: 300, Limit: defaultLimit, CountTotal: false}
+	request = types.NewQueryAllBalancesRequest(addr1, pageReq)
+	res, err = queryClient.AllBalances(gocontext.Background(), request)
+	require.NoError(t, err)
+	require.LessOrEqual(t, res.Balances.Len(), 0)
+	require.Nil(t, res.Res.NextKey)
+}
+
+func ExamplePaginate() {
+	app, ctx := setupTest()
+	queryHelper := baseapp.NewQueryServerTestHelper(ctx)
+	types.RegisterQueryServer(queryHelper, app.BankKeeper)
+	queryClient := types.NewQueryClient(queryHelper)
+
+	var balances sdk.Coins
+
+	for i := 0; i < 2; i++ {
+		denom := fmt.Sprintf("foo%ddenom", i)
+		balances = append(balances, sdk.NewInt64Coin(denom, 100))
+	}
+
+	addr1 := sdk.AccAddress([]byte("addr1"))
+	acc1 := app.AccountKeeper.NewAccountWithAddress(ctx, addr1)
+	app.AccountKeeper.SetAccount(ctx, acc1)
+	err := app.BankKeeper.SetBalances(ctx, addr1, balances)
+	if err != nil {
+		fmt.Println(err)
+	}
+	pageReq := &query.PageRequest{Key: nil, Limit: 1, CountTotal: true}
+	request := types.NewQueryAllBalancesRequest(addr1, pageReq)
+	res, _ := queryClient.AllBalances(gocontext.Background(), request)
+	fmt.Println(res)
```

Comment:
I think ExamplePaginate should directly deal with Paginate and should not use grpc call. That way, it will be purely Paginate example. @alessio what do you think?

```suggestion

	app, ctx := setupTest()

	var balances sdk.Coins

	for i := 0; i < 2; i++ {
		denom := fmt.Sprintf("foo%ddenom", i)
		balances = append(balances, sdk.NewInt64Coin(denom, 100))
	}

	addr1 := sdk.AccAddress([]byte("addr1"))
	acc1 := app.AccountKeeper.NewAccountWithAddress(ctx, addr1)
	app.AccountKeeper.SetAccount(ctx, acc1)
	err := app.BankKeeper.SetBalances(ctx, addr1, balances)
	if err != nil {
		fmt.Println(err)
	}

	// Paginate example
	pageReq := &query.PageRequest{Key: nil, Limit: 1, CountTotal: true}
	request := types.NewQueryAllBalancesRequest(addr1, pageReq)
	balResult := sdk.NewCoins()
	store := ctx.KVStore(app.GetKey(authtypes.StoreKey))
	balancesStore := prefix.NewStore(store, types.BalancesPrefix)
	accountStore := prefix.NewStore(balancesStore, addr1.Bytes())

	res, err := query.Paginate(accountStore, request.Req, func(key []byte, value []byte) error {
		var tempRes sdk.Coin
		err := app.Codec().UnmarshalBinaryBare(value, &tempRes)
		if err != nil {
			return err
		}

		balResult = append(balResult, tempRes)
		return nil
	})

	if err != nil { // should return no error
		fmt.Println(err)
	}

	fmt.Println(&types.QueryAllBalancesResponse{Balances: balResult, Res: res})
```

Reply:
Yeah, this gives more context to a new user. Will update

Reply:
Good point @anilCSE, I agree